### PR TITLE
Return non-empty arrays where possible

### PR DESCRIPTION
**Pull request by thewilkybarkid**
For #1404.

I've started with the `Task` cases for feedback; there are a lot more modules that can also be updated.

### CODE REVIEW

File: docs/modules/Task.ts.md [hunk 1]
```diff
@@ -595,7 +595,9 @@ Equivalent to `ReadonlyArray#sequence(ApplicativePar)`.
 **Signature**
 
 ```ts
-export declare const sequenceArray: <A>(arr: readonly Task<A>[]) => Task<readonly A[]>
+export declare const sequenceArray: <A, B>(
+  arr: readonly Task<A>[] & { readonly 0?: Task<B> | undefined }
+) => Task<Task<B> extends Task<A> ? ReadonlyNonEmptyArray<A> : readonly A[]>
```
Comment (thewilkybarkid) [inline docs/modules/Task.ts.md:600]: This signature is probably confusing to read...

Comment (gcanti) [reply docs/modules/Task.ts.md:600]: Yeah, I prefer two distinct functions with clean signatures

**Comment by thewilkybarkid** [comment]
I think the type problems are the same as https://github.com/gcanti/fp-ts/pull/1436#discussion_r599383444. I'll have another look then see what happens when applying it to v3, which I suspect would work.

If that's the case, I'll probably reopen this with a nicer PR to v3, then decide whether it's worth adding some temporary functions to v2.

**Comment by thewilkybarkid** [comment]
Just seen that the separate functions exist in v3: https://github.com/gcanti/fp-ts/blob/cf9f33bd512382c04d9b388c870e8018da48ffeb/src/Task.ts#L414-L433

Guess backporting is preferred?

**Comment by gcanti** [comment]
@thewilkybarkid yes, I'm still not sure about the names though

**Comment by thewilkybarkid** [comment]
👍  I'll close this then.

**Pull request closed without merging**